File: app/core/ConditionSeverityLoader.php
<?php
/**
 * Canonical medical condition → NON_URGENT | URGENT | EMERGENCY from CSV.
 * Used by ClinicalTriageEngine so final triage is CSV-driven, not LLM-driven.
 */

final class ConditionSeverityLoader
{
    /** @var array<string, array<string, mixed>>|null */
    private static ?array $index = null;

    public static function clearCache(): void
    {
        self::$index = null;
    }

    /** @return array<string, array<string, mixed>> */
    public static function index(): array
    {
        if (self::$index !== null) {
            return self::$index;
        }

        self::$index = [];
        // Curated runtime registry only (ICD overlay is for validation / offline tooling).
        self::loadFile(BASE_PATH . '/data/nlp/condition_triage_severity.csv', true);

        return self::$index;
    }

    /**
     * @param list<string> $terms
     * @return array<string, mixed>|null
     */
    public static function lookup(array $terms): ?array
    {
        $index = self::index();
        $rank = ['NON_URGENT' => 0, 'URGENT' => 1, 'EMERGENCY' => 2];
        $best = null;

        foreach ($terms as $term) {
            $key = self::normalize((string) $term);
            if ($key === '') {
                continue;
            }
            $meta = $index[$key] ?? null;
            if ($meta === null) {
                continue;
            }
            if ($best === null || $rank[$meta['severity_level']] > $rank[$best['severity_level']]) {
                $best = $meta;
            }
        }

        return $best;
    }

    private static function loadFile(string $path, bool $prefer): void
    {
        if (!is_readable($path)) {
            return;
        }
        $handle = fopen($path, 'r');
        if ($handle === false) {
            return;
        }
        $header = fgetcsv($handle);
        if ($header === false) {
            fclose($handle);
            return;
        }
        $header = array_map(static fn ($h) => strtolower(trim((string) $h)), $header);
        $source = basename($path);

        while (($row = fgetcsv($handle)) !== false) {
            $data = [];
            foreach ($header as $i => $col) {
                $data[$col] = trim((string) ($row[$i] ?? ''));
            }
            $level = self::canonicalize($data['severity_level'] ?? '');
            $name = (string) ($data['medical_condition'] ?? $data['condition_name'] ?? '');
            if ($level === '' || $name === '') {
                continue;
            }
            $meta = [
                'medical_condition'   => $name,
                'severity_level'       => $level,
                'urgency_score'       => (int) ($data['urgency_score'] ?? 0) ?: self::defaultScore($level),
                'emergency_flag'      => in_array(strtolower((string) ($data['emergency_flag'] ?? '0')), ['1', 'true', 'yes'], true),
                'recommended_action'  => (string) ($data['recommended_action'] ?? ''),
                'provider_required'   => in_array(strtolower((string) ($data['provider_required'] ?? '0')), ['1', 'true', 'yes'], true),
                'hospital_referral'   => in_array(strtolower((string) ($data['hospital_referral'] ?? '0')), ['1', 'true', 'yes'], true),
                'source'              => $source,
            ];
            self::add($name, $meta, $prefer);
            foreach (explode(';', (string) ($data['synonyms'] ?? '')) as $syn) {
                self::add($syn, $meta, false);
            }
            self::add((string) ($data['hiligaynon_term'] ?? ''), $meta, false);
            foreach (explode(';', (string) ($data['keywords'] ?? '')) as $kw) {
                self::add($kw, $meta, false);
            }
        }
        fclose($handle);
    }

    /** @param array<string, mixed> $meta */
    private static function add(string $key, array $meta, bool $prefer): void
    {
        $k = self::normalize($key);
        if ($k === '') {
            return;
        }
        if (isset(self::$index[$k]) && !$prefer) {
            return;
        }
        self::$index[$k] = $meta;
    }

    private static function normalize(string $text): string
    {
        $text = mb_strtolower(trim($text));
        $text = preg_replace('/[^a-z0-9\s\-]/u', ' ', $text) ?? $text;

        return trim(preg_replace('/\s+/u', ' ', $text) ?? $text);
    }

    private static function canonicalize(string $raw): string
    {
        $v = strtoupper(str_replace(['-', ' '], '_', trim($raw)));
        if (in_array($v, ['NON_URGENT', 'URGENT', 'EMERGENCY'], true)) {
            return $v;
        }
        $map = [
            'non_urgent' => 'NON_URGENT',
            'routine'    => 'NON_URGENT',
            'low'        => 'NON_URGENT',
            'urgent'     => 'URGENT',
            'high'       => 'URGENT',
            'emergency'  => 'EMERGENCY',
            'critical'   => 'EMERGENCY',
        ];
        $low = strtolower(str_replace(['-', ' '], '_', trim($raw)));

        return $map[$low] ?? '';
    }

    private static function defaultScore(string $level): int
    {
        return match ($level) {
            'EMERGENCY' => 90,
            'URGENT'    => 55,
            default     => 20,
        };
    }
}

File: ai_service/condition_severity_loader.py
"""Load canonical condition triage severity from CSV (not LLM)."""

from __future__ import annotations

import csv
import re
from functools import lru_cache
from pathlib import Path
from typing import Any

_NLP = Path(__file__).resolve().parent.parent / "data" / "nlp"

_STORAGE_TO_CLASS = {
    "non_urgent": "NON_URGENT",
    "non-urgent": "NON_URGENT",
    "routine": "NON_URGENT",
    "low": "NON_URGENT",
    "urgent": "URGENT",
    "high": "URGENT",
    "emergency": "EMERGENCY",
    "critical": "EMERGENCY",
    "non_urgent".upper(): "NON_URGENT",
    "URGENT": "URGENT",
    "EMERGENCY": "EMERGENCY",
    "NON_URGENT": "NON_URGENT",
}


def _norm(text: str) -> str:
    text = (text or "").lower().strip()
    text = re.sub(r"[^a-z0-9\s\-]", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def _canon(raw: str) -> str:
    v = (raw or "").strip().upper().replace("-", "_").replace(" ", "_")
    if v in {"NON_URGENT", "URGENT", "EMERGENCY"}:
        return v
    low = (raw or "").strip().lower().replace("-", "_").replace(" ", "_")
    return _STORAGE_TO_CLASS.get(low, "")


@lru_cache(maxsize=1)
def condition_severity_index() -> dict[str, dict[str, Any]]:
    """
    Index English condition/synonym/Hiligaynon → severity metadata.
    Runtime uses curated condition_triage_severity.csv.
    ICD overlay is opt-in via MEDCONNECT_LOAD_ICD_TRIAGE_OVERLAY=1 (large).
    """
    import os

    index: dict[str, dict[str, Any]] = {}

    def add(key: str, meta: dict[str, Any], prefer: bool = False) -> None:
        k = _norm(key)
        if not k:
            return
        if k in index and not prefer:
            return
        index[k] = meta

    registry = _NLP / "condition_triage_severity.csv"
    if registry.is_file():
        with registry.open(encoding="utf-8", newline="") as f:
            for row in csv.DictReader(f):
                level = _canon(row.get("severity_level") or "")
                if not level:
                    continue
                meta = {
                    "medical_condition": (row.get("medical_condition") or "").strip(),
                    "severity_level": level,
                    "urgency_score": int(row.get("urgency_score") or 0) or (
                        20 if level == "NON_URGENT" else 55 if level == "URGENT" else 90
                    ),
                    "emergency_flag": (row.get("emergency_flag") or "0").strip() in {"1", "true", "yes"},
                    "recommended_action": (row.get("recommended_action") or "").strip(),
                    "provider_required": (row.get("provider_required") or "0").strip() in {"1", "true", "yes"},
                    "hospital_referral": (row.get("hospital_referral") or "0").strip() in {"1", "true", "yes"},
                    "source": "condition_triage_severity.csv",
                }
                add(meta["medical_condition"], meta, prefer=True)
                for part in (row.get("synonyms") or "").split(";"):
                    add(part, meta)
                add(row.get("hiligaynon_term") or "", meta)
                for part in (row.get("keywords") or "").split(";"):
                    add(part, meta)

    if os.environ.get("MEDCONNECT_LOAD_ICD_TRIAGE_OVERLAY", "").strip() in {"1", "true", "yes"}:
        overlay = _NLP / "medical_conditions_triage_overlay.csv"
        if overlay.is_file():
            with overlay.open(encoding="utf-8", newline="") as f:
                for row in csv.DictReader(f):
                    level = _canon(row.get("severity_level") or "")
                    name = (row.get("medical_condition") or "").strip()
                    if not level or not name:
                        continue
                    meta = {
                        "medical_condition": name,
                        "severity_level": level,
                        "urgency_score": int(row.get("urgency_score") or 0) or (
                            20 if level == "NON_URGENT" else 55 if level == "URGENT" else 90
                        ),
                        "emergency_flag": (row.get("emergency_flag") or "0").strip() in {"1", "true", "yes"},
                        "recommended_action": (row.get("recommended_action") or "").strip(),
                        "provider_required": (row.get("provider_required") or "0").strip() in {"1", "true", "yes"},
                        "hospital_referral": (row.get("hospital_referral") or "0").strip() in {"1", "true", "yes"},
                        "source": "medical_conditions_triage_overlay.csv",
                    }
                    add(name, meta, prefer=False)

    return index


def lookup_condition_severity(*terms: str) -> dict[str, Any] | None:
    """Return highest-urgency exact match among terms (CSV-backed, no LLM)."""
    index = condition_severity_index()
    rank = {"NON_URGENT": 0, "URGENT": 1, "EMERGENCY": 2}
    best: dict[str, Any] | None = None
    for term in terms:
        meta = index.get(_norm(term))
        if not meta:
            continue
        if best is None or rank[meta["severity_level"]] > rank[best["severity_level"]]:
            best = meta
    return best


def clear_cache() -> None:
    condition_severity_index.cache_clear()
